dali/pipeline/executor/queue_policy.h

```diff
@@ -0,0 +1,331 @@
+    if (HasPreviousStage(stage)) {
+      int previous_stage = static_cast<int>(PreviousStage(stage));
+      std::unique_lock<std::mutex> ready_previous_lock(stage_ready_mutex_[previous_stage]);
+      stage_ready_cv_[previous_stage].wait(ready_previous_lock, [previous_stage, this]() {
+        return !stage_ready_[previous_stage].empty() || exec_error_;
+      });
+      if (exec_error_) {
+        return QueueIdxs{-1};
+      }
+      // result[static_cast<OpType>(previous_stage)] = stage_ready_[previous_stage].front();
+      // We fill the information about all the previous stages herew
+      result = stage_ready_[previous_stage].front();
+      stage_ready_[previous_stage].pop();
+      // We are the only ones waiting for the lock, so we do not try to wake anyone
+    }
+    // There always is a current stage
+    {
+      std::unique_lock<std::mutex> free_current_lock(stage_free_mutex_[current_stage]);
+      stage_free_cv_[current_stage].wait(free_current_lock, [current_stage, this]() {
+        return !stage_free_[current_stage].empty() || exec_error_;
+      });
+      if (exec_error_) {
+        return QueueIdxs{-1};
+      }
+      // We add info about current stage
+      result[stage] = stage_free_[current_stage].front();
+      stage_free_[current_stage].pop();
+    }
+    return result;
+  }
+
+  void ReleaseIdxs(OpType stage, QueueIdxs idxs) {
+    int current_stage = static_cast<int>(stage);
+    // We have a special case for Support ops - they are set free by a GPU stage,
+    // during QueueOutputIdxs
+    if (stage != OpType::CPU) {
+      if (HasPreviousStage(stage)) {
+        ReleaseStageIdx(PreviousStage(stage), idxs);
+      }
+    }
+    {
+      std::unique_lock<std::mutex> ready_current_lock(stage_ready_mutex_[current_stage]);
+      // stage_ready_[current_stage].push(idxs[stage]);
+      // Store the idxs up to the point of stage that we processed
+      stage_ready_[current_stage].push(idxs);
+    }
+    stage_ready_cv_[current_stage].notify_one();
+  }
+
+  void QueueOutputIdxs(QueueIdxs idxs) {
+    std::unique_lock<std::mutex> ready_output_lock(ready_output_mutex_);
+    ready_output_queue_.push({idxs[OpType::MIXED], idxs[OpType::GPU]});
+    ready_output_lock.unlock();
+    ready_output_cv_.notify_all();
+
+    // In case of GPU we release also the Support Op
+    ReleaseStageIdx(OpType::SUPPORT, idxs);
+  }
+
+  OutputIdxs UseOutputIdxs() {
+    // Block until the work for a batch has been issued.
+    // Move the queue id from ready to in_use
+    std::unique_lock<std::mutex> ready_lock(ready_output_mutex_);
+    ready_output_cv_.wait(ready_lock,
+                          [this]() { return !ready_output_queue_.empty() || exec_error_; });
+    if (exec_error_) {
+      return OutputIdxs{-1, -1};
+    }
+    auto output_idx = ready_output_queue_.front();
+    ready_output_queue_.pop();
+    in_use_queue_.push(output_idx);  // TODO(klecki) -this may cause some problems!!!
```

Comment:
This is actually modified only in synchronous calls from python, so we should be fine. 
Still, left a todo, in case we allow for some asynchronous access here.